core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,695 @@
+                    'class': int. Associated feedback class with program.
+                    'tokens': list(str). Token list of program.
+                    'k_gram_hashes': list(int). K-gram hash values of program.
+                } for each program in training_data.
+            }
+    """
+    for program_id in training_data:
+        training_data[program_id]['k_gram_hashes'] = (
+            winnowing.k_gram_hash_generator(
+                training_data[program_id]['tokens'], token_to_id, K))
+    return training_data
+
+
+def generate_program_fingerprints(training_data, T, K):
+    """Generate document fingerprints for all programs in entire dataset.
+
+    Args:
+        training_data: dict. A dictionary containing training data. Structure of
+            training_data is as follows,
+                {
+                    ID: {
+                        'source': str. Source code of program.
+                        'class': int. Associated feedback class with program.
+                        'tokens': list(str). Token list of program.
+                        'k_gram_hashes': list(int). K-gram hash values of
+                            program.
+                    } for each program in training_data.
+                }
+        T: int. Model parameter 'T' of winnowing.
+        K: int. Model parameter 'K' of winnowing.
+
+    Returns:
+        dict. A dictionary containing training data with additional key
+        'fingerprint' for each program which stores generated fingerprint
+        of program. Structure of dict is as follows,
+            {
+                ID: {
+                    'source': str. Source code of program.
+                    'class': int. Associated feedback class with program.
+                    'tokens': list(str). Token list of program.
+                    'k_gram_hashes': list(int). K-gram hash values of program.
+                    'fingerprint': list(int). Extracted fingerprint of program.
+                } for each program in training_data.
+            }
+    """
+    window_size = T - K + 1
+    for program_id in training_data:
+        training_data[program_id]['fingerprint'] = (
+            winnowing.get_fingerprint_from_hashes(
+                training_data[program_id]['k_gram_hashes'], window_size))
+    return training_data
+
+
+def calc_jaccard_index(set_a, set_b):
+    """Calculate jaccard's coefficient for two sets set_a and set_b.
+
+    Jaccard index of two set is equal to:
+        (no. of elements in intersection of two sets)
+        _____________________________________________
+        (no. of elements in union of two sets)
+
+    Args:
+        set_a: list(int). First set.
+        set_b: list(int). Second set.
+
+    Returns:
+        float. Jaccard index of two sets.
+
+    """
+    small_set = set_a[:] if len(set_a) < len(set_b) else set_b[:]
+    union_set = set_b[:] if len(set_a) < len(set_b) else set_a[:]
+    for elem in union_set:
```

Comment:
Why are you doing this? You already have union and intersection in Python [defined](https://docs.python.org/2/library/sets.html#example) for sets.